source/extensions/filters/network/dubbo_proxy/decoder.h

```diff
@@ -13,38 +14,137 @@ namespace Extensions {
+  void setCurrentState(ProtocolState state) { state_ = state; }
+
+private:
+  struct DecoderStatus {
+    DecoderStatus() = default;
+    DecoderStatus(ProtocolState next_state) : next_state_(next_state), filter_status_{} {};
+    DecoderStatus(ProtocolState next_state, Network::FilterStatus filter_status)
+        : next_state_(next_state), filter_status_(filter_status){};
+
+    ProtocolState next_state_;
+    absl::optional<Network::FilterStatus> filter_status_;
+  };
+
+  // These functions map directly to the matching ProtocolState values. Each returns the next state
+  // or ProtocolState::WaitForData if more data is required.
+  DecoderStatus onTransportBegin(Buffer::Instance& buffer, Protocol::Context& context);
+  DecoderStatus onTransportEnd();
+  DecoderStatus onTransferHeaderTo(Buffer::Instance& buffer, size_t length);
+  DecoderStatus onTransferBodyTo(Buffer::Instance& buffer, int32_t length);
+  DecoderStatus onMessageBegin();
+  DecoderStatus onMessageEnd(Buffer::Instance& buffer, int32_t message_size);
+
+  // handleState delegates to the appropriate method based on state_.
+  DecoderStatus handleState(Buffer::Instance& buffer);
+
+  Protocol& protocol_;
+  Deserializer& deserializer_;
+  MessageMetadataSharedPtr metadata_;
+  DecoderCallbacks& decoder_callbacks_;
+
+  ProtocolState state_;
+  Protocol::Context context_;
+
+  DecoderEventHandler* handler_;
+};
+
+typedef std::unique_ptr<DecoderStateMachine> DecoderStateMachinePtr;
+
 /**
  * Decoder encapsulates a configured and ProtocolPtr and SerializationPtr.
  */
 class Decoder : public Logger::Loggable<Logger::Id::dubbo> {
 public:
-  Decoder(ProtocolPtr&& protocol, DeserializerPtr&& deserializer,
-          DecoderCallbacks& decoder_callbacks);
+  Decoder(Protocol& protocol, Deserializer& deserializer, DecoderCallbacks* decoder_callbacks);
 
   /**
    * Drains data from the given buffer
    *
    * @param data a Buffer containing Dubbo protocol data
    * @throw EnvoyException on Dubbo protocol errors
    */
-  void onData(Buffer::Instance& data);
+  Network::FilterStatus onData(Buffer::Instance& data, bool& buffer_underflow);
 
-  const Deserializer& serializer() { return *deserializer_; }
-  const Protocol& protocol() { return *protocol_; }
+  const Deserializer& serializer() { return deserializer_; }
+  const Protocol& protocol() { return protocol_; }
 
 private:
-  DeserializerPtr deserializer_;
-  ProtocolPtr protocol_;
-  bool decode_ended_ = false;
-  Protocol::Context context_;
-  DecoderCallbacks& decoder_callbacks_;
+  struct ActiveRequest {
+    ActiveRequest(DecoderEventHandler& handler) : handler_(handler) {}
+
+    DecoderEventHandler& handler_;
+  };
+  typedef std::unique_ptr<ActiveRequest> ActiveRequestPtr;
+
+  void start();
+  void complete();
+
+  MessageMetadataSharedPtr metadata_;
+  Deserializer& deserializer_;
+  Protocol& protocol_;
+  DecoderStateMachinePtr state_machine_;
+  bool decode_started_ = false;
+  DecoderCallbacks* decoder_callbacks_;
```

Comment:
If decoder_callbacks_ is not nullptr, it should be referenced here.

Reply:
done